delegate-app/HydraAuction/Delegate/Server.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RankNTypes #-}
+
+module HydraAuction.Delegate.Server (
+  -- * Delegate server types
+  DelegateServerConfig (..),
+  DelegateServerLog (..),
+  DelegateError (..),
+
+  -- * tracing types
+
+  -- ** tracing transformer
+  TracerT (..),
+  runWithTracer,
+  runWithTracer',
+  mapTracerT,
+  askTracer,
+
+  -- ** tracing type class
+  MonadTracer (..),
+
+  -- ** delegate tracing
+  DelegateTracer,
+
+  -- * wai extras
+  ServerAppT,
+) where
+
+-- Prelude imports
+import Hydra.Prelude (ReaderT (runReaderT), withReaderT)
+import Prelude
+
+-- Haskell imports
+import Control.Tracer (Contravariant (contramap), Tracer, natTracer, traceWith)
+import GHC.Generics (Generic)
+import Prettyprinter (Pretty (pretty), line, viaShow)
+
+-- Cardano imports
+import Hydra.Network (IP, PortNumber)
+
+-- Hydra auction imports
+
+import Control.Monad.RWS
+import HydraAuction.Delegate.Interface (DelegateResponse, FrontendRequest)
+import Network.WebSockets (PendingConnection)
+
+-- | The config for the delegate server
+data DelegateServerConfig = DelegateServerConfig
+  { dlgt'host :: IP
+  -- ^ the host of the delegate server
+  , dlgt'port :: PortNumber
+  -- ^ the port number the delegate server receives input at
+  , dlgt'tick :: Int
+  -- ^ the amount of milliseconds, the thread should wait
+  }
+
+data DelegateServerLog
+  = Started PortNumber
+  | FrontendConnected
+  | FrontendInput FrontendRequest
+  | DelegateOutput DelegateResponse
+  | DelegateError DelegateError
+  deriving stock (Eq, Show, Generic)
+
+instance Pretty DelegateServerLog where
+  pretty = \case
+    Started port -> "Started Server at Port" <> pretty (show port)
```

Comment:
All other logs just use show so this may look strange in CLI.

Reply:
this is what IOG uses largely, I think it's quite nice because we can just in a modular way structure our logs and adjust the error messages whenever we want to. I don't see why it would look strange in CLI, it would just be formatted a bit differently, I think it's nicer than adhoc strings in the code.

Reply:
this also serves as documentation